Foundation/JSONSerialization.swift

```diff
@@ -812,6 +812,35 @@ private struct JSONReader {
         0x2E, 0x2D, 0x2B, 0x45, 0x65, // . - + E e
     ]
     func parseNumber(_ input: Index, options opt: JSONSerialization.ReadingOptions) throws -> (Any, Index)? {
+       func castInt64(_ value: Int64) -> Any {
+            if Int.bitWidth == 64 {
+                return Int(value)
+            } else {
+                if let result = Int(exactly: value) {
+                    return result
+                }
+
+                return value
+            }
+        }
+
+        func castUInt64(_ value: UInt64) -> Any {
+            if let result = Int(exactly: value) {
+                return result
+            }
+
+            if Int.bitWidth == 64 {
+                return UInt(value)
+            } else {
+                if let result = Int64(exactly: value) {
+                    return result
+                }
+
+                return value
+            }
+        }
+
+
         func parseTypedNumber(_ address: UnsafePointer<UInt8>, count: Int) -> (Any, IndexDistance)? {
             let temp_buffer_size = 64
```

Comment:
Anybody knows why 64 char length hardcoded here?

Reply:
I'm guessing the idea here is that we won't be able to represent any number >64 digits long faithfully; this isn't necessarily a correct way to do this, though.

Reply:
I am trying to parse `String(repeating: "9", count: 165)` on Darwin and it parsed as 1.0e165 which is much better than 1.0e64
But 166 failed with an exception, which works fine in js in chrome.
```
Domain=NSCocoaErrorDomain Code=3840 "Number wound up as NaN around character 1."
```

Reply:
What OS version are you running?

Reply:
macOS 10.12.6 Xcode 9.2

Reply:
So yes, on macOS Sierra (10.12) and previous, you'll see the older behavior of a reduced range of values which parse as a `double` and more numbers which fall into the case parsing an `NSDecimalNumber` (which has a more limited range).

Reply:
Ok long floating point numbers works even worse then I think. Looks like another big issue for another PR

This code fails on swift 4.0.2 on Ubuntu 16.04
```
import Foundation

let json = "[\(Int.max - 100).5]".data(using: .utf8)!
let data = try! JSONSerialization.jsonObject(with: json)

print(data)
```